### DISCUSSION TITLE
Can we discuss what is left to do in order to stabilize the subscription API?

### DISCUSSION THREAD

**Discussion by ghost** [trpc/trpc#866] [General]
tRPC is exactly what I always wanted to make/use since I started using TypeScript. The moment I watched the demo gif in the README I knew I found web dev heaven :smile:

My project makes use of subscriptions but the API is currently marked as experimental. I think there should be a task list of what needs to be changed/improved. We could gather feedback here or create a pinned issue. So far I haven't found any problems with it.

**Comment by KATT** [comment]
Oi João!

The subscriptions are pretty stable as far as I know - the only reason I've said they aren't is because I haven't used them in production myself yet.

**Comment by ianmartorell** [comment]
Hey @KATT! I'm considering implementing tRPC in our express backend, and a simultaneous move to serverless is looking very tempting. The only piece I can't quite figure out how to fit in the puzzle is subscriptions. It seems like the current implementation uses WebSockets, and that seems like it's not a viable option / would be too expensive when using a serverless service like Vercel. Is this the case in your opinion? Would you deploy a backend using the current WebSockets Subscriptions model to Vercel, or would that be more suited to be hosted on a more traditional server environment like Heroku?

**Reply by KATT** [reply]
You can't do real-time and serverless unless you're doing longpollling which will be insanely expensive. Some people use akka instead of out subscription layer for their real-time stuff which seems to work well.

Render.com works well with subscriptions and is easier than heroku, has autoscaling etc

**Reply by ianmartorell** [reply]
That's what I thought, thanks for your input! As another alternative for someone else who might read this, it seems Vercel is [recommending](https://vercel.com/support/articles/do-vercel-serverless-functions-support-websocket-connections) a pub-sub approach with a service like pusher.com.

**Reply by jlalmes** [reply]
AWS API Gateway does offer a cost-effective serverless WebSocket solution with Lambdas, albeit I have never tried this setup with tRPC subscriptions (https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/). +1 for Render :+1:

**Reply by KATT** [reply]
_(FYI - AWS WebSockets doesn't really work with tRPC's subscriptions, but you can use it the same way you could use pusher)_

**Reply by mmkal** [reply]
@KATT curious what you found that didn’t work - we have a ticket to try this out (not sure when we’ll get to it). We are all in on AWS for fun HIPAA reasons so likely would not be able to use Render.

**Reply by KATT** [reply]
How AWS handles websockets in their api gateway is a bit peculiar, I don't remember enough of the implementation to say what/why, but as from how I remember it its not really a "server", it's more like you have different event handlers like onConnect/onMessage which triggers lambdas and then you have to have your own storage thing to keep track of the connected clients that you can then use to pass messages to a specific connection. 

Tl;Dr - fundamentally very different mental model from a classic node websocket server. 

Would be happy to have a subscriptions that allowed for this, but I likely won't champion this myself as it's big, hairy and I have no use of it as of writing.

**Reply by KATT** [reply]
You can also run trpc in ECS/EKS on AWS and use a websockets server

**Reply by misterjoshua** [reply]
For clarity: Any system that wants to use AWS API Gateway's WebSockets for this sort of pub-sub must track connections and connection subscriptions in out-of-process, persistent storage instead of relying on in-process memory like a classic node WebSocket server. When broadcasting to subscribers, any AWS API Gateway implementation must look up all subscribed connections in persistent storage and post to each.

A third-party implementation looks like it's trying to tackle this problem, here: https://github.com/purple-technology/serverless-trpc-subscriptions

**Comment by cyrfer** [comment]
I am also interested in TRPC/Lambda support for AWS Websocket events.

**Comment by silberistgold** [comment]
I am also interested in websocket support for trpc in lambdas.

Looking at the example [here](https://trpc.io/docs/subscriptions) I understand that `observable` is stateful at the moment because we need access to the `emit` function when an event happens. That clashes with the stateless nature of lambda/serverless. I guess we would need a stateless version of `observable` which allows to keep track of subscriptions in arbitrary storage.
